### PR TITLE
implemented tags/node_tags stats downloading

### LINKED ISSUE #5412: add CSV tag and node_tag stats downloads
Hi, this is a [first-timers-only issue](https://code.publiclab.org/#r=all). This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. 

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you 💝

If you have contributed before, **consider leaving this one for someone new**, and looking through our general [help wanted](https://github.com/publiclab/plots2/labels/help-wanted) issues. Thanks!

### 🤔 What you will need to know.

Nothing. This issue is meant to welcome you to Open Source :) We are happy to walk you through the process.

### 📋 Step by Step

- [ ] 🙋 **Claim this issue**: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!

- [ ] 📝 **Update** the file [app/controllers/stats_controller.rb](https://github.com/publiclab/plots2/blob/master/app/controllers/stats_controller.rb) in the `plots2` repository (press the little pen Icon) and edit the line as shown below. 

[See this page](https://code.publiclab.org/#r=all) for some help in taking your first steps!

Below is a "diff" showing in red (and a `-`) which lines to remove, and in green (and a `+`) which lines to add:

```diff
@@ -100,6 +100,16 @@ def answers
     format(data, 'answers')
   end
 
+  def tag
+    data = Tag.select(:tid, :name, :parent, :count).all
+    format(data, 'tag')
+  end
+
+  def node_tag
+    data = NodeTag.select(:tid, :nid, :uid).where(date: start.to_i...fin.to_i).all
+    format(data, 'node_tag')
+  end
+
   def comments
     data = Comment.where(status: 1, timestamp: start.to_i...fin.to_i).all
     format(data, 'comment')
```

- [ ] 💾 **Commit** your changes

- [ ] 🔀 **Start a Pull Request**. There are two ways how you can start a pull request:

1. If you are familiar with the terminal or would like to learn it, [here is a great tutorial](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github) on how to send a pull request using the terminal.

2. You can also [edit files directly in your browser](https://help.github.com/articles/editing-files-in-your-repository/) and open a pull request from there. 

- [ ] 🏁 **Done** Ask in comments for a review :)

### Please keep us updated

💬⏰ - We encourage contributors to be respectful to the community and provide an update **within a week** of claiming a first-timers-only issue. We're happy to keep it assigned to you as long as you need if you update us with a request for more time or help, but if we don't see any activity a week after you claim it we may reassign it to give someone else a chance. Thank you in advance! 

If this happens to you, don't sweat it! Grab another open issue.

### Is someone else already working on this?

🔗- We encourage contributors to link to the original issue in their pull request so all users can easily see if someone's already started on it. 

👥- **If someone seems stuck, offer them some help!** Otherwise, [take a look at some other issues you can help with](https://code.publiclab.org/#r=all). Thanks!

### 🤔❓ Questions?

Leave a comment below!

### PR DESCRIPTION
**Pull request by skilfullycurled**
Fixes #5412 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] (see note below) screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

If tests do fail, click on the red `X` to learn why by reading the logs.

### CODE REVIEW

**Comment by skilfullycurled** [comment]
@publiclab/reviewers: checks are all passed, and ready for review!  

One thing: since I did not make the changes on a separately installed version of plots, I can't give a screenshot of what the drop down looks like with the [changes](https://github.com/publiclab/plots2/pull/5459/commits/f37ae75ccb542ea95258433c34c5674ce9f5b664).  The UI should still be the same, just with an additional selection.

Please let me know what else I can do/improve upon in the PR process!

File: app/controllers/stats_controller.rb [hunk 1]
```diff
@@ -105,6 +105,16 @@ def comments
     format(data, 'comment')
   end
 
+  def tag
```
Comment (gauravano) [inline app/controllers/stats_controller.rb]:
```suggestion
  def tags
```

File: app/controllers/stats_controller.rb [hunk 2]
```diff
@@ -105,6 +105,16 @@ def comments
...
+    data = Tag.select(:tid, :name, :parent, :count).all
+    format(data, 'tag')
+  end
+
+  def node_tag
```
Comment (gauravano) [inline app/controllers/stats_controller.rb]:
```suggestion
  def node_tags
```

File: config/routes.rb [hunk 3]
```diff
@@ -261,6 +261,9 @@
   get 'stats/questions/:start/:end' => 'stats#questions'
   get 'stats/answers' => 'stats#answers'
   get 'stats/answers/:start/:end' => 'stats#answers'
+  get 'stats/tags' => 'stats/tags'
```
Comment (gauravano) [inline config/routes.rb]:
```suggestion
  get 'stats/tags' => 'stats#tags'
```

File: app/views/stats/_range.html.erb [hunk 4]
```diff
@@ -93,6 +99,11 @@
                 <li>
                     <%= link_to "Questions Answered", stats_answers_path(format: 'csv', start: params[:start], end: params[:end]) %>
                 </li>
+                    <%= link_to "List of Tags", stats_tags_path(format: 'csv') %>
```
Comment (gauravano) [inline app/views/stats/_range.html.erb]:
```suggestion
                <li>
                    <%= link_to "List of Tags", stats_tags_path(format: 'csv') %>
```

Comment (gauravano) [reply app/views/stats/_range.html.erb]: Opening tag needs to be opened here

Comment (gauravano) [reply app/views/stats/_range.html.erb]: Also, @skilfullycurled there's an extra `<li>` at https://github.com/publiclab/plots2/pull/5459/files#diff-e8d970f814aa8ace0c4be3778174c22bR92. This is unrelated to your code, but I think we can remove that too. Thanks!

File: app/controllers/stats_controller.rb (hunk 2 above)
Comment (skilfullycurled) [reply app/controllers/stats_controller.rb]: I'm laying [this one](https://github.com/publiclab/plots2/issues/5412#issuecomment-481769938) on @jywarren  : P  No, no, jk, jk.  It was amazing that he wrote it out for me.  And I should have caught that. I've been looking a lot at the models and controllers lately and there are ones that can be referenced above.  : )

**Comment by skilfullycurled** [comment]
@gauravano, thank you so much!  This has been an awesome learning experience.  Besides the process in general, I had no idea there would be features like "suggestions" which make this really easy when you make mistakes.  

Speaking of which, all of these are good reminders that I need to do a better job of re-reading my code.  It's a bit harder for me to catch mistakes because Ruby is new to me, but I also know that in a different language, I would have done a bit closer of a reading.  I'll definitely need to install the website somewhere so I can have the feedback of "events not working as expected".

Hopefully the CI will finish soon!

**Comment by skilfullycurled** [comment]
Okay, removed the other tag per your [comment](https://github.com/publiclab/plots2/pull/5459/files#diff-e8d970f814aa8ace0c4be3778174c22bL86) above.

Let the checks begin (again).

File: app/controllers/stats_controller.rb (hunk 2 above)
Comment (skilfullycurled) [reply app/controllers/stats_controller.rb]: Whoops! Forgot to commit it.

**Comment by skilfullycurled** [comment]
@gauravano, all checks passed!

PS: any feedback you can give me about communication regarding the review process would be welcome.  Specifically:

1) when checks pass the first time, should I @reply pl/reviewers as I did, or do people from the team go through all of the pull requests, say once a week or something so it's unneccessary?

2) when checks pass after the first review and changes are made/accepted, should I @reply the person who reviewed the request (in this case, you), do you get some notification since now you're a reviewer?

Thanks, I just want to make sure that I'm not spamming people.

**Comment by gauravano** [comment]
Hi @skilfullycurled,  

> when checks pass the first time, should I @reply pl/reviewers as I did, or do people from the team go through all of the pull requests, say once a week or something so it's unneccessary?

You can mention the reviewers after your work is ready for review or if you face any problem while solving the issue. Like, you mentioned @publiclab/reviewers after doing all the work, but at the same time you made clear that you haven't set up the project locally, which is really helpful because now reviewer will test your PR on their machine. 

You should mention the reviewers as only some people/reviewers try to go through all the PRs and others view a PR only when they get the notification.

> when checks pass after the first review and changes are made/accepted, should I @reply the person who reviewed the request (in this case, you), do you get some notification since now you're a reviewer?

As sometimes that same person might not be available so you can mention that person and reviewers both.

Thanks!

**Comment by gauravano** [comment]
I can review this PR again in the evening, but if any other @publiclab/plots2-reviewers is free, they'll leave a review. Thanks!

**Comment by skilfullycurled** [comment]
Thanks @gauravano.  I appreciate the permission, I don't want to seem like I'm bugging people, I imagine everyone's inboxes are full!

**Comment by cesswairimu** [comment]
So I just pulled the changes locally
here is the UI looking great :tada: 
curl

The `download as json` part is working fine. But the `download as csv` throws this error

curl2

Also noticed it does not filter according the range selected but rather return all the data. Will add suggestions for both of these fixes below.

**Comment by cesswairimu** [comment]
Oh my bad just remembered we don't have a timestamp for the tags table.

**Comment by skilfullycurled** [comment]
Thanks @cesswairimu!  I added that.  Just waiting for the CI check to pass.  Let me know if you have a "style" preference for adding extends.  

For example:
```
class NodeTag < ApplicationRecord
  extend RawStats
  self.table_name = 'community_tags'
```
vs. 
```
class NodeTag < ApplicationRecord

  extend RawStats
  self.table_name = 'community_tags'
```
vs.
```
class NodeTag < ApplicationRecord
  extend RawStats

  self.table_name = 'community_tags'
```

**Comment by skilfullycurled** [comment]
All tests passed! Ready for (re)review @publiclab/plots2-reviewers!

**Comment by cesswairimu** [comment]
@skilfullycurled I have no preference, your changes are looking awesome.
Good job :tada:

**Comment by jywarren** [comment]
This is awesome, @skilfullycurled ! It's not 100% necessary to do such a big one as your FTO in a single step, but awesome that you did! 

Now that we're this far, i had one more request -- could you change line 8 of this file:

`test/functional/stats_controller_test.rb`

to this? 

`    @stats =  [:notes, :comments, :users, :wikis, :questions, :answers, :tags, :node_tags]`

That'll add tests for your two new methods, and protect them against future changes!

**Comment by skilfullycurled** [comment]
Sure thing!

**Comment by skilfullycurled** [comment]
Okay.  All set, will return when checks have passed.

**Comment by skilfullycurled** [comment]
Okay.  All set @publiclab/plots2-reviewers!  There's definitely a bit of a "watched kettle never boils" situation while waiting for that CI.

**Comment by jywarren** [comment]
Fantastic work!!!! Thanks so much @skilfullycurled !!!

**Comment by skilfullycurled** [comment]
\o/

@jywarren, would it be possible to have permissions on https://stable.publiclab.org/ so I can test it out per the bot's request...?

**Comment by jywarren** [comment]
Sure!

**Comment by skilfullycurled** [comment]
Oh, forgot, I wanted to thank everyone.  As many other people have mentioned, it was a great experience.  Thanks for all the support, positive feedback, and celebrations.  On to the next issue!

**Pull request merged**